src/util.jl

```diff
@@ -36,76 +52,171 @@ function stim_tableau(stim_sim::Py)::Matrix{Int}
+    stim_sim = state.simulator
+    @disp_time "\tcurrent_inverse_tableau: " tb = stim_sim.current_inverse_tableau()
+    @disp_time "\tinverse: " tb = tb.inverse()
+    oldlen = state.qubits
+    state.qubits = qubits = length(tb)
+    svec = state.stabilizers
+    pt = t0 = time_ns()
+    cnt = 0
+    @disp_time "\tupdate stabilizers: \n" begin
+    if isempty(svec)
+        for i in 1:qubits
+            ps = tb.z_output(i-1)
+            xs, zs = ps.to_numpy()
+            xv = pyconvert(BitVector, xs)
+            zv = pyconvert(BitVector, zs)
+            push!(svec, Stabilizer(xv, zv, pyconvert(Number, ps.sign) == -1 ? 2 : 0))
+            if debug_out && (cnt += 1) > 9999
+                tn = time_ns() ; out_time(i, tn-pt, tn-t0, qubits) ; pt = tn
+                cnt = 0
+            end
+        end
+    else
+        qubits == oldlen || error("Mismatch qubits $qubits != $oldlen, $(length(svec))")
+        for i in 1:qubits
+            ps = tb.z_output(i-1)
+            xs, zs = ps.to_numpy()
+            stab = svec[i]
+            stab.X = pyconvert(BitVector, xs)
+            stab.Z = pyconvert(BitVector, zs)
+            stab.phase = pyconvert(Number, ps.sign) == -1 ? 2 : 0
+            if debug_out && (cnt += 1) > 9999
+                tn = time_ns() ; out_time(i, tn-pt, tn-t0, qubits) ; pt = tn
+                cnt = 0
+            end
+        end
+    end
+    debug_out && (tn = time_ns() ; out_time(cnt, tn-pt, tn-t0))
+    end
 
-    # create a new state temporarily
-    inbetween_state = tableau_to_state(tableau)
+    # mark it as updated
+    state.is_updated = true
+end
 
-    # update the initial state
-    state.qubits = inbetween_state.qubits
-    state.stabilizers = deepcopy(inbetween_state.stabilizers)
+"""
+    rand(StabilizerState, qubits::Int)
+
+    Return a random (valid) stabilizer state with the given number of qubits
+"""
+function Base.rand(::Type{StabilizerState}, qubits::Int)
+    tableau = stim.Tableau.random(qubits).inverse()
+    state = StabilizerState(qubits)
+    state.simulator.set_inverse_tableau(tableau)
+    svec = state.stabilizers
+    (_, _, z2x, z2z, _, z_signs) = tableau.to_numpy()
+    signs = pyconvert(BitVector, z_signs)
+    for i in 1:qubits
+        xv = pyconvert(BitVector, z2x[i-1])
+        zv = pyconvert(BitVector, z2z[i-1])
+        push!(svec, Stabilizer(xv, zv, signs[i]<<1))
+    end
 
     # mark it as updated
     state.is_updated = true
+    return state
+end
+
+function _is_symmetric(svec::Vector{Stabilizer})
+    qubits = length(svec)
+    for i = 1:qubits-1
+        sz = svec[i].Z
+        for j = i+1:qubits
+            sz[j] == svec[j].Z[i] || return false
+        end
+    end
+    return true
 end
 
+export to_stabilizer_graph
 """
-    to_graph(state)
+    to_stabilizer_graph(state)
 
 Convert a state to its adjacency graph state equivalent under local operations.
+
+Returns a vector of Stabilizers, and a vector of operations performed
```

Comment:
[comment] Naming is a bit confusing for me here. You say it "returns a vector of stabilizers", but then the name of the function says `to_stabilizer_graph`? So is it a graph or not really?

Reply:
The Z bits of the stabilizers make a graph.  It returns the vector of stabilizers instead of having to allocate a new vector to store just Z BitVectors (for performance reasons)

Reply:
If you can think of a better name, I'm all ears!

Reply:
`graph_as_stabilizer_vector` maybe?

Reply:
`get_stabilizer_vectore_representing_graph` would be a mouthful 😆 
`graph_as_stabilizer_vector` sounds good. Does it need to be public or can we make it private?

Reply:
Ok!